python/ray/tune/manager.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import requests
+import json
+import threading
+from http.server import HTTPServer, BaseHTTPRequestHandler
+from six.moves.queue import Queue
+
+import ray
+from ray.tune.error import TuneError, TuneManagerError
+from ray.tune.variant_generator import generate_trials
+
+@ray.remote
+def _send_request(path, payload):
+    response = requests.get(path, data=payload)
+    parsed = response.json()
+    return parsed
+
+
+class ExpManager(object):
+    STOP = "STOP"
+    ADD = "ADD"
+    GET_LIST = "GET_LIST"
+    GET_TRIAL = "GET_TRIAL"
+
+    def __init__(self, tune_address):
+        # TODO(rliaw): Better to specify address and port forward
+        self._tune_address = tune_address
+        self._path = "http://{}".format(tune_address)
+
+    def get_all_trials(self, nowait=False):
+        """Returns a list of all trials (trial_id, config, status)"""
+        return self._get_response(
+            {"command": ExpManager.GET_LIST}, nowait)
+
+    def get_trial(self, trial_id, nowait=False):
+        """Returns the last result for queried trial"""
+        return self._get_response(
+            {"command": ExpManager.GET_TRIAL,
+             "trial_id": trial_id}, nowait)
+
+    def add_trial(self, name, trial_spec, nowait=False):
+        """Adds a trial of `name` with configurations"""
+        # TODO(rliaw): have better way of specifying a new trial
+        return self._get_response(
+            {"command": ExpManager.ADD,
+             "name": name,
+             "spec": trial_spec}, nowait)
+
+    def stop_trial(self, trial_id, nowait=False):
+        return self._get_response(
+            {"command": ExpManager.STOP,
+             "trial_id": trial_id}, nowait)
+
+    def _get_response(self, data, nowait=False):
+        payload = json.dumps(data).encode() # don't know if needed
+        response_future = _send_request.remote(self._path, payload)
+        if nowait:
+            return response_future
+        else:
+            return ray.get(response_future)
+
+def QueueHandler(in_queue, out_queue):
+    class Handler(BaseHTTPRequestHandler):
+
+        def do_GET(self):
+            content_len = int(self.headers.get('Content-Length'), 0)
+            raw_body = self.rfile.read(content_len)
+            parsed_input = json.loads(raw_body.decode())
+            in_queue.put(parsed_input)
+            status, response = out_queue.get()
+            if status:
+                self.send_response(200)
+            else:
+                self.send_response(400)
+            self.end_headers()
+            self.wfile.write(json.dumps(
+                response).encode())
+    return Handler
+
+
+class TuneManager(threading.Thread):
```

Comment:
note: would be clearer what's going on if the classes were named *Server / *Client